Optimizations to mikktspace.c

Sluggernot opened the pull request:
Created in response to: https://github.com/godotengine/godot/issues/93587 
Large glb file (778 MB) would hang, crash or load after extremely long time. This first set of optimizations focuses on sending SVec3 objects by reference instead of by value. Local tests in debug mode caused one iteration of GenerateSharedVerticesIndexList to go from 552ms to 470ms, on average. Unsure of the performance gains on release mode.

Sluggernot committed:
Optimizations to mikktspace.c

Created in response to: https://github.com/godotengine/godot/issues/93587
Large glb file (778 MB) would hang, crash or load after extremely long time. This first set of optimizations focuses on sending SVec3 objects by reference instead of by value. Local tests in debug mode caused one iteration of GenerateSharedVerticesIndexList to go from 552ms to 470ms, on average. Unsure of the performance gains on release mode.

Sluggernot committed:
Optimizations to mikktspace.c

Created in response to: https://github.com/godotengine/godot/issues/93587
Large glb file (778 MB) would hang, crash or load after extremely long time. This first set of optimizations focuses on sending SVec3 objects by reference instead of by value. Local tests in debug mode caused one iteration of GenerateSharedVerticesIndexList to go from 552ms to 470ms, on average. Unsure of the performance gains on release mode.

Sluggernot committed:
Next commit, whitespace corrections

freakin whitespace

Sluggernot committed:
Merge branch 'master' of https://github.com/Sluggernot/godot

Sluggernot commented:
First time using Github like this. Hopefully this isn't too bad.

AThousandShips commented:
These changes must be added as a patch file in `thirdparty/misc/patches` as they are third party sources and can be updated upstream, losing our changes

Mickeon commented:
If it can be done it would be worth considering reporting this issue to the third-party, as well.

AThousandShips commented:
The patch would help a lot with that as well

akien-mga commented:
For the record, upstream is now https://github.com/mmikk/MikkTSpace and no longer just a page on the (old, and now 404'ing) Blender wiki.

But before proposing this upstream, this should definitely be benchmarked in a production build, debug build times aren't super useful aside from seeing the impact on debugging usability for engine devs.

Sluggernot commented:
Is there a standard benchmarking system of any kind? Any info would be helpful. 
The last changes to that repo is 4 years ago but I'll attempt a pull request.
Can I propose to replace these two files with a cpp implementation?

fire commented:
Can yoy try to salvage https://github.com/godotengine/godot/pull/83648

akien-mga commented:
> Is there a standard benchmarking system of any kind? Any info would be helpful. The last changes to that repo is 4 years ago but I'll attempt a pull request. Can I propose to replace these two files with a cpp implementation?

This is reference code for the implementation used by Blender, which is in C.
So I don't think porting to C++ would be welcome, though a separate (duplicated) reference implementation in C++ may be ok, you can ask if they're interested.

This is also not a library that changed at all in the past 10 years, so it's not a problem if we hard fork it to optimize it for Godot's needs.

Sluggernot commented:
https://github.com/mmikk/MikkTSpace/issues/5  a 5 month old Issue where someone is asking for 2 '=' to be added to correct some sorting. (I will say I haven't vetted the request but it hasn't been commented on either.) Confidence is low. But, I'm not here to talk trash about a legacy/third-party lib. It was amazing until we want to load larger files.

What's the real solution? 
My findings (which can be flawed) are showing the major bottlenecks being sqrt calculation with Length calls and calls to VScale. VScale has nothing going on computationally. The issue must be copy elision. Creating an SVec3 and returning a copy. C doesn't allow me to make overloaded constructors for structs (as far as I can see. I'm a C++ guy way more than C). I'm looking at other ways to get around this but open to suggestions.

akien-mga commented:
I think it's find modifying this lib for our needs, it's a single file and not one that gets updated frequently. If porting to C++ lets us use features that improve performance in Godot's context, then I don't see a reason not to do it.

Sluggernot committed:
Initial suggestion for speeding up glb load times

There is still lots of room for improvement but with my local tests I am down to 322ms from 552ms before my other proposed changes. The new major bottlenecks are QuickSortEdges and mikktGetPosition.
Issues: I am temporarily unable to build a release build, locally. I will look into this to get real world numbers.
I don't know how copyright works in this instance. I would happily keep the original copyright and hand these changes over.
I left a large comment in surface_tool.h about how to handle this side-change to the use of these new files without removing the old mikktspace.

Sluggernot commented:
I have so much to learn about git. 
I have created a new commit that isn't a new pull request... It is an initial C++ implementation of mikktspace 
There are several options to consider before accepting this change. I will try to get real world timings for these changes but I wanted to get some opinions from you guys before I dive further and spend time on something that may be impossible to actually complete. 
While writing this I see I have a lot of fails. Looks like linting errors.

Sluggernot committed:
Trying to fix the linting error I received.

The error is ambiguous and oddly seemed to point the order in which I included two files. I need more verbosity or I just missed the description.

fire commented:
I would recommend getting timing 2-3 times on the https://github.com/godotengine/godot/issues/93587.

@Calinou has a benchmarking server we can use too!

fire commented:
I've had good experiences writing doctest unit tests for these things complicated things, it's not a requirement for merging, but wanted to mention.

Calinou commented:
> I would recommend getting timing 2-3 times on the https://github.com/godotengine/godot/issues/93587.

Note that you can get Godot to print the time taken to import each resource by starting it with `--verbose` from the terminal, then looking at terminal output after each file is done importing.

fire commented:
- [ ] benchmarked in a production build
- [ ] benchmarked using the test cases (like linked issues)
- [ ] optional unit tests

Sluggernot committed:
Over engineered change for improving load times of glb files.

I was in search of every possible way to improve load times (of glb files in my case) and created this monstrosity of a set of changes. Do not allow this change to go through. I just wanted to preserve this set of changes for personal reasons.

Sluggernot commented:
Alright, change incoming. I think I will have you reject the change and you will see why. 
I did a ton of digging into the loading code, using Visual Studio perf tools to get timings at different sections of code, mostly encompassing editor_file_system.cpp _reimport_file. I looked at the timings for import compared to reimport of my glb files and found them to be very similar. I have a ~34MB glb file called chess_set that was taking ~18 seconds to load. Perfect amount of time to iterate but get good numbers for a test, imo. 
My performance tests led me to a couple of major functions taking almost the entire time, in a file called simplifier.cpp. rankEdgeCollapses and two functions both called quadricError. I decided these functions could use some more inline elements and not allocate as much memory since they are called so many times. These changes resulted in measurable but fairly insignificant results. 
I created new versions that removed an if check and a for loop. The variable attribute_count was a hard coded '3' so if we are going to have branches, it may be worth creating functions that don't have those. (Foreshadowing: It is not) SUCCESS! My new timings I checked over and over took the load time from 18 seconds to 4 seconds!
Raw Note: EditorFileSystem: "res://chess_set.glb" import took 4833 ms. WOAH! The changes to rankEdgeCollapses3 and quadricError3 were freaking massive?!
So, I diligently checked other files and yes we can achieve about a 75% increase! But, something bugged me with one change. I shouldn't have checked it. The results were too heartbreaking after all of my work. 
In the super insane loop that is rankEdgeCollapses, I had taken the variable c out from the loop and declared it before the loop. A nice little optimization so we create that Collapse& object 1 time. How much did this particular optimization account for? About 90% of the entire set of all other optimizations, I believe. 

tl;dr
Feel free to go to simplifier.cpp line 933 and just make it look like this:
```	Collapse& c = collapses[0];
	for (size_t i = 0; i < collapse_count; ++i) {
		 c = collapses[i];
```
You should see the import of glb files (and I don't even know what else) massively improve. :( 
I'll get that change together shortly. I have to collect myself.

Sluggernot committed:
Revert "Over engineered change for improving load times of glb files."

This reverts commit 87c90773fd39c50128c71ad8e7640d2753e19fce.

Sluggernot committed:
Revert "Trying to fix the linting error I received."

This reverts commit 3f7b156fc5b484daac5974d9272e1cf26ae527d0.

Sluggernot committed:
Revert "Initial suggestion for speeding up glb load times"

This reverts commit f883e6eb420bf7c6587844c47107005dbdb2cc15.

Sluggernot committed:
Revert "Next commit, whitespace corrections"

This reverts commit 27a6fb3d2b800d451338fe58aa3388d75c5a7be7.

Sluggernot committed:
Revert "Optimizations to mikktspace.c"

This reverts commit 65276837cafe9796644a8fd33c62f6bd3f399ee4.

Sluggernot commented:
https://github.com/zeux/meshoptimizer/pull/713

Hashing out some potential results with the owner of a potentially huge fix.

fire commented:
I'm going to try running some timing tests on my folder of 3d assets.

Sluggernot commented:
Are you reverting the reverts I did? The changes, imo, should be tested after Zeux change. I am fairly certain it would be beneficial on a file like that 700+ MB glb file called scene that we're all using. The initial file processing should be improved. (And with the Zeux change, wont take so long to test)

fire commented:
4.3 has released, this is one of the pull requests I wanted to shepherd to completion.
